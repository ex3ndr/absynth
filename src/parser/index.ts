// export { AbsynthLexer } from './lexer';
// export { AbsynthParser } from './parser';

// export type ASTNodeType = 'number' | 'string' | 'declaration' | 'assignment' | 'program' | 'operation' | 'statements';

interface ASTBaseNode {
    first_column: number;
    first_line: number;
    last_column: number;
    last_line: number;
}

export interface ASTGraphqlType extends ASTBaseNode {
    type: 'type';
    name: ASTId;
}

export interface ASTNumber extends ASTBaseNode {
    type: 'number';
    value: number;
}

export interface ASTString extends ASTBaseNode {
    type: 'string';
    value: string;
}

export interface ASTBoolean extends ASTBaseNode {
    type: 'boolean';
    value: boolean;
}

export interface ASTDeclaration extends ASTBaseNode {
    type: 'declaration';
    name: string;
}

export interface ASTReference extends ASTBaseNode {
    type: 'reference';
    name: string;
}

export interface ASTAssignment extends ASTBaseNode {
    type: 'assignment';
    left: ASTDeclaration;
    right: ASTExpression;
}

export interface ASTProgram extends ASTBaseNode {
    type: 'program';
    value: ASTNode[];
}

export interface ASTOperation extends ASTBaseNode {
    type: 'operation';
    op: string;
    left: ASTExpression;
    right: ASTExpression;
}

export interface ASTStatements extends ASTBaseNode {
    type: 'statements';
    statements: ASTNode[];
}

export interface ASTCall extends ASTBaseNode {
    type: 'call';
    name: string;
    arguments: ASTExpression[];
}

export interface ASTModel extends ASTBaseNode {
    type: 'model';
    name: ASTId;
    modifiers: ASTModifier[];
    fields: ASTModelField[];
}

export interface ASTId extends ASTBaseNode {
    type: 'id';
    name: string;
}

export interface ASTModelField extends ASTBaseNode {
    type: 'field';
    name: ASTId;
    field_type: ASTType;
    modifiers: ASTModifier[];
}

export interface ASTType extends ASTBaseNode {
    type: 'type_primitive' | 'type_reference';
    name: string;
    id: ASTId;
    required: boolean;
    resolvedType: ASTModel | ASTEnum;
}

export interface ASTModifier extends ASTBaseNode {
    type: 'modifier';
    name: string;
}

export interface ASTEnum extends ASTBaseNode {
    type: 'enum';
    name: ASTId;
    values: ASTString[];
}

export type ASTExpression = ASTOperation | ASTNumber | ASTString | ASTReference | ASTBoolean | ASTCall;

export type ASTNode = ASTNumber | ASTString | ASTDeclaration | ASTAssignment | ASTProgram | ASTOperation | ASTStatements | ASTCall | ASTModel | ASTModifier | ASTType | ASTId | ASTModelField | ASTEnum | ASTBoolean | ASTGraphqlType;