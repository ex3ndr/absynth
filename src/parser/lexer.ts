import Lexer from 'lex';

export class AbsynthLexer {
    private lexer: Lexer;
    constructor(table: string[][]) {
        this.lexer = new Lexer();

        for (let rules of table) {
            this.lexer.addRule(new RegExp(rules[0]), function (lexeme: string) {
                this.yytext = lexeme;
                if (rules[1] === '') {
                    return undefined;
                }
                return rules[1];
            });
        }
        this.lexer.addRule(/$/, function () {
            return 'EOF';
        });
    }

    setInput = (src: string) => {
        this.lexer.setInput(src);
    }

    lexerInstance() {
        return this.lexer;
    }

    lex(source: string) {
        this.lexer.setInput(source);
        let res: string[] = [];
        while (this.lexer.index < source.length) {
            res.push(this.lexer.lex());
        }
        return res;
    }

    lexDiagnostics(source: string): number {
        this.lexer.setInput(source);
        while (this.lexer.index < source.length) {
            try {
                this.lexer.lex();
            } catch (e) {
                console.warn(e);
                return this.lexer.index;
            }
        }
        return -1;
    }
}