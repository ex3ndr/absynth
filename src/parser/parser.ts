import { Parser } from 'jison-gho';
import Lexer from 'lex';
import { ASTProgram } from '.';

interface ParserException {
    hash: {
        recoverable: boolean;
        lexer: Lexer;
        loc?: {
            first_line: number,
            first_column: number,
            last_line: number,
            last_column: number,
            range: number[]
        }
    };
}

export class AbsynthParser {
    private parser: Parser;

    constructor(lexer: string[][], parsing: { [key: string]: string[][] }) {
        this.parser = new Parser({
            lex: {
                rules: lexer.map((v) => [v[0], v[1] !== '' ? 'return \'' + v[1] + '\';' : '']),
            },
            ebnf: {
                ...parsing,
            }
        });
    }

    parse(source: string): ASTProgram {
        return this.parser.parse(source);
    }

    parseDiagnostics(source: string): { loc: number[], msg: string } | undefined {
        try {
            this.parser.parse(source);
        } catch (e) {
            // console.warn(e);
            let ex = e as ParserException;
            if (ex.hash.loc) {
                console.log('Lexer error');
                console.log(ex.hash.loc);
                // console.log(ex.hash);
                return { loc: ex.hash.loc.range, msg: e.message };
            } else {
                // console.log((ex as any).hash);
            }
            // console.warn(ex.hash.lexer.index);
            // console.warn(ex.hash.lexer !== this.lexer.lexerInstance());
            // return ex.hash.lexer.index;
        }
        return undefined;
    }
}