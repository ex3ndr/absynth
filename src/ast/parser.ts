import { AbsynthLexer, lexerTable } from './lexer';
import { Parser } from 'jison-gho';
import Lexer from 'lex';
import { ASTProgram } from '.';

interface ParserException {
    hash: {
        recoverable: boolean;
        lexer: Lexer;
        loc?: {
            first_line: number,
            first_column: number,
            last_line: number,
            last_column: number,
            range: number[]
        }
    };
}

function location(start?: number | string, end?: number | string) {
    let s = start || 1;
    let e = end || 1;
    return ', ' +
        'first_column: @' + s + '.first_column, ' +
        'first_line: @' + s + '.first_line, ' +
        'last_column: @' + e + '.last_column, ' +
        'last_line: @' + e + '.last_line';
}

export class AbsynthParser {
    private lexer: AbsynthLexer;
    private parser: Parser;

    constructor() {
        this.lexer = new AbsynthLexer();
        this.parser = new Parser({
            lex: {
                rules: lexerTable.map((v) => [v[0], v[1] !== '' ? 'return \'' + v[1] + '\';' : '']),
            },
            ebnf: {

                //
                // Program Root
                //

                'program': [['root_declarations EOF', 'return {type: \'program\', value: $1' + location() + '};']],

                //
                // Root Sections
                //

                'root_declarations': [
                    ['root_declarations root_declaration', '$$ = [...$1, $2]'],
                    ['root_declaration', '$$ = [$1]']
                ],

                'root_declaration': [
                    ['model', '$$ = $1'],
                    ['statements_block', '$$ = $1'],
                    ['statement', '$$ = $1'],
                ],

                //
                // Model Definitions
                //

                'model': [
                    ['modifiers T_MODEL id model_fields_block', '$$ = {type:\'model\', modifiers: $1, name: $3, fields: $4' + location(2, 4) + '};']
                ],

                'model_fields_block': [
                    ['T_BRACE_OPEN model_fields T_BRACE_CLOSE', '$$ = $2'],
                    ['T_BRACE_OPEN T_BRACE_CLOSE', '$$ = [];'],
                    //   ['model_block model_field'],
                    //   ['model_field']
                ],

                'model_fields': [
                    ['model_fields model_field', '$$ = [...$1, $2];'],
                    ['model_field', '$$ = [$1];']
                ],

                'model_field': [
                    ['modifiers T_FIELD id T_COLON type', '$$ = {type:\'field\', modifiers: $1, name: $3, field_type: $5' + location(2, 5) + '}']
                ],

                'modifiers': [
                    ['modifiers modifier', '$$ = [...$1, $2]'],
                    // ['T_MODIFIER', '$$ = {type:\'modifiers\', keywords: yytext' + location(1) + '}'],
                    ['', '$$ = []']
                ],

                'modifier': [
                    ['T_MODIFIER', '$$ = {type:\'modifier\', name: yytext' + location(1) + '}']
                ],

                'type': [
                    ['T_PRIMITIVE', '$$ = {type:\'type_primitive\', required: false, name: yytext' + location(1) + '}'],
                    ['T_PRIMITIVE T_EXCLAMATION', '$$ = {type:\'type_primitive\', required: true, name: $1' + location(1) + '}'],
                    ['id', '$$ = {type:\'type_reference\', required: false, id: $1' + location(1) + '}'],
                    ['id T_EXCLAMATION', '$$ = {type:\'type_reference\', required: true, id: $1' + location(1) + '}']
                ],

                //
                // Statements
                //

                'statements_block': [
                    ['T_BRACE_OPEN statements T_BRACE_CLOSE', '$$ = $2'],
                    ['T_BRACE_OPEN T_BRACE_CLOSE', '$$ = {type:\'statements\', statements: []' + location(1, 2) + '};'],
                ],

                'statements': [
                    ['statements statement', '$$ = {type:\'statements\', statements: [...$1.statements, $2]' + location(1, 2) + '};'],
                    ['statement', '$$ = {type:\'statements\', statements: [$1]' + location(1) + '};'],
                ],

                'statement': [
                    ['call T_SEMICOLON?', '$$ = $1;'],
                    ['assignment T_SEMICOLON?', '$$ = $1;'],
                    ['T_RETURN expression T_SEMICOLON?', '$$ = {type:\'return\', expression: $2' + location(1, 2) + ' }'],
                    ['declaration_enum T_SEMICOLON?', '$$ = $1;']
                ],

                //
                // Variable declaration
                //

                'declaration_variable': [['T_LET IDENTIFIER', '$$ = {type: \'declaration\', name: yytext ' + location(1) + '};']],
                'assignment': [['declaration_variable EQUALS expression', '$$ = {type: \'assignment\', left: $1, right: $3' + location(1, 3) + '};']],

                //
                // Enum
                //

                'declaration_enum': [['T_ENUM id EQUALS (constant_string[v] (T_BAR constant_string[v])*)[values]', '$$ = {type: \'enum\', name: $2, values: [$values[0], ...$values[1].map((v)=>v[1])]' + location(1, 'values') + '};']],

                //
                // Global function call
                //

                'call': [
                    ['IDENTIFIER T_BRACKET_OPEN arguments T_BRACKET_CLOSE', '$$ = {type:\'call\', name: $1, arguments: $3' + location(1, 4) + '};'],
                ],
                'arguments': [
                    ['expression T_COMMA arguments', '$$ = [$1, ...$3];'],
                    ['expression', '$$ = [$1];'],
                    ['', '$$ = [];']
                ],

                //
                // Expressions of all kind. Anything that eventually resolved to a value.
                //

                'expression': [
                    ['expression_additive', '$$ = $1;'],
                ],
                'expression_additive': [
                    ['expression_multiple T_PLUS expression_additive', '$$ = {type:\'operation\', op: \'+\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_multiple T_MINUS expression_additive', '$$ = {type:\'operation\', op: \'-\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_multiple', '$$ = $1;']
                ],
                'expression_multiple': [
                    ['constant T_MULT expression_multiple', '$$ = {type:\'operation\', op: \'*\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['constant T_DIV expression_multiple', '$$ = {type:\'operation\', op: \'/\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_terminal', '$$ = $1;']
                ],
                'expression_terminal': [
                    ['call', '$$=$1'],
                    ['constant', '$$ = $1']
                ],
                'constant': [
                    ['IDENTIFIER', '$$ = {type: \'reference\', name: yytext' + location(1) + '};'],
                    ['constant_number', '$$ = $1;'],
                    ['constant_string', '$$ = $1;'],
                ],
                'id': [
                    ['IDENTIFIER', '$$ = {type: \'id\', name: yytext' + location(1) + '};'],
                ],
                'constant_number': [
                    ['NUMBER_HEX', '$$ = {type: \'number\', value: parseInt(yytext)' + location(1) + '};'],
                    ['NUMBER', '$$ = {type: \'number\', value: parseInt(yytext)' + location(1) + '};']
                ],
                'constant_string': [
                    ['STRING', '$$ = {type: \'string\', value: yytext.substring(1, yytext.length-1)' + location(1) + '};']
                ]
            }
        });
        // this.parser.lexer = this.lexer.lexerInstance();
    }

    parse(source: string): ASTProgram {
        return this.parser.parse(source);
    }

    parseDiagnostics(source: string): { loc: number[], msg: string } | undefined {
        try {
            this.parser.parse(source);
        } catch (e) {
            // console.warn(e);
            let ex = e as ParserException;
            if (ex.hash.loc) {
                console.log('Lexer error');
                console.log(ex.hash.loc);
                // console.log(ex.hash);
                return { loc: ex.hash.loc.range, msg: e.message };
            } else {
                // console.log((ex as any).hash);
            }
            // console.warn(ex.hash.lexer.index);
            // console.warn(ex.hash.lexer !== this.lexer.lexerInstance());
            // return ex.hash.lexer.index;
        }
        return undefined;
    }
}