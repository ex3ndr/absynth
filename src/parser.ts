import { AbsynthLexer } from './lexer';
import { Parser } from 'jison-gho';

export interface ASTNode {
    type: string;
}

export interface ASTNumber extends ASTNode {
    type: 'number';
    value: number;
}

export interface ASTString extends ASTNode {
    type: 'string';
    value: string;
}

export interface ASTDeclaration extends ASTNode {
    type: 'declaration';
    name: string;
}

export interface ASTAssignment extends ASTNode {
    type: 'assignment';
    left: ASTDeclaration;
    right: ASTNode;
}

export interface ASTProgram extends ASTNode {
    type: 'program';
    value: ASTNode;
}

export interface ASTOperation extends ASTNode {
    type: 'operation';
    op: 'string';
    left: ASTNode;
    right: ASTNode;
}

export class AbsynthParser {
    private lexer: AbsynthLexer;
    private parser: Parser;

    constructor() {
        this.lexer = new AbsynthLexer();
        this.parser = new Parser({
            bnf: {
                'program': [['statements EOF', 'return {type: \'program\', value: $1};']],

                'statements': [
                    ['statements T_SEMICOLON statement', '$$ = {type:\'statements\', statements: [...$1.statements, $3]};'],
                    ['statements T_SEMICOLON', '$$ = $1;'],
                    ['statement', '$$ = {type:\'statements\', statements: [$1]};']
                ],

                'statement': [
                    ['call', '$$ = $1;'],
                    ['assignment', '$$ = $1;']
                ],

                'call': [
                    ['IDENTIFIER T_BRACKET_OPEN arguments T_BRACKET_CLOSE', '$$ = {type:\'call\', name: $1, arguments: $3};']
                ],

                'arguments': [
                    ['expression T_COMMA arguments', '$$ = [$1, ...$3];'],
                    ['expression', '$$ = [$1];'],
                    ['', '$$ = [];']
                ],

                'assignment': [['declaration EQUALS expression', '$$ = {type: \'assignment\', left: $1, right: $3};']],
                'declaration': [['T_LET IDENTIFIER', '$$ = {type: \'declaration\', name: yytext};']],

                'expression': [
                    ['expression_additive', '$$ = $1;']
                ],
                'expression_additive': [
                    ['expression_multiple T_PLUS expression_additive', '$$ = {type:\'operation\', op: \'+\', left: $1, right: $3};'],
                    ['expression_multiple T_MINUS expression_additive', '$$ = {type:\'operation\', op: \'-\', left: $1, right: $3};'],
                    ['expression_multiple', '$$ = $1;']
                ],
                'expression_multiple': [
                    ['constant T_MULT expression_multiple', '$$ = {type:\'operation\', op: \'*\', left: $1, right: $3};'],
                    ['constant T_DIV expression_multiple', '$$ = {type:\'operation\', op: \'/\', left: $1, right: $3};'],
                    ['constant', '$$ = $1;']
                ],
                'constant': [
                    ['constant_number', '$$ = $1;'],
                    ['constant_string', '$$ = $1;'],
                ],
                'constant_number': [
                    ['NUMBER_HEX', '$$ = {type: \'number\', value: parseInt(yytext)};'],
                    ['NUMBER', '$$ = {type: \'number\', value: parseInt(yytext)};']
                ],
                'constant_string': [
                    ['STRING', '$$ = {type: \'string\', value: yytext};']
                ]
            }
        });
        this.parser.lexer = this.lexer.lexerInstance();
    }

    parse(source: string): ASTProgram {
        return this.parser.parse(source);
    }

    parseDiagnostics(source: string): number {
        try {
            this.parser.parse(source);
        } catch (e) {
            return this.lexer.lexerInstance().index;
        }
        return -1;
    }
}