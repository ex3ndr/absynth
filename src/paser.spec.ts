import { AbsynthParser } from './parser';
describe('Parser', () => {
    it('parse simple assignment expression with strings', () => {
        let parser = new AbsynthParser();
        let res = parser.parse('let i = \'123\'');
        expect(res).toMatchSnapshot();
    });
    it('parse simple assignment expression with calculus', () => {
        let parser = new AbsynthParser();
        let res = parser.parse('let i = 123 + 1/2');
        expect(res).toMatchSnapshot();
    });
    it('should parse statements', () => {
        let parser = new AbsynthParser();
        let res = parser.parse('let i = 123 + 1/2; let j = 0;');
        expect(res).toMatchSnapshot();
    });
    it('should parse calls', () => {
        let parser = new AbsynthParser();
        let res = parser.parse('println()');
        expect(res).toMatchSnapshot();
        res = parser.parse('println(123,3444)');
        expect(res).toMatchSnapshot();
    });
});