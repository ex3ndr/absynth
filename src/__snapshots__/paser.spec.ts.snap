// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parse simple assignment expression with calculus 1`] = `
Object {
  "type": "program",
  "value": Object {
    "statements": Array [
      Object {
        "left": Object {
          "name": "i",
          "type": "declaration",
        },
        "right": Object {
          "left": Object {
            "type": "number",
            "value": 123,
          },
          "op": "+",
          "right": Object {
            "left": Object {
              "type": "number",
              "value": 1,
            },
            "op": "/",
            "right": Object {
              "type": "number",
              "value": 2,
            },
            "type": "operation",
          },
          "type": "operation",
        },
        "type": "assignment",
      },
    ],
    "type": "statements",
  },
}
`;

exports[`Parser parse simple assignment expression with strings 1`] = `
Object {
  "type": "program",
  "value": Object {
    "statements": Array [
      Object {
        "left": Object {
          "name": "i",
          "type": "declaration",
        },
        "right": Object {
          "type": "string",
          "value": "123",
        },
        "type": "assignment",
      },
    ],
    "type": "statements",
  },
}
`;

exports[`Parser should parse calls 1`] = `
Object {
  "type": "program",
  "value": Object {
    "statements": Array [
      Object {
        "arguments": Array [],
        "name": "println",
        "type": "call",
      },
    ],
    "type": "statements",
  },
}
`;

exports[`Parser should parse calls 2`] = `
Object {
  "type": "program",
  "value": Object {
    "statements": Array [
      Object {
        "arguments": Array [
          Object {
            "type": "number",
            "value": 123,
          },
          Object {
            "type": "number",
            "value": 3444,
          },
        ],
        "name": "println",
        "type": "call",
      },
    ],
    "type": "statements",
  },
}
`;

exports[`Parser should parse statements 1`] = `
Object {
  "type": "program",
  "value": Object {
    "statements": Array [
      Object {
        "left": Object {
          "name": "i",
          "type": "declaration",
        },
        "right": Object {
          "left": Object {
            "type": "number",
            "value": 123,
          },
          "op": "+",
          "right": Object {
            "left": Object {
              "type": "number",
              "value": 1,
            },
            "op": "/",
            "right": Object {
              "type": "number",
              "value": 2,
            },
            "type": "operation",
          },
          "type": "operation",
        },
        "type": "assignment",
      },
      Object {
        "left": Object {
          "name": "j",
          "type": "declaration",
        },
        "right": Object {
          "type": "number",
          "value": 0,
        },
        "type": "assignment",
      },
    ],
    "type": "statements",
  },
}
`;
